go/plugins/compat_oai/openai/generate.go

```diff
@@ -0,0 +1,132 @@
+package openai
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/firebase/genkit/go/ai"
+	"github.com/openai/openai-go"
+)
+
+// ModelGenerator handles OpenAI generation requests
+type ModelGenerator struct {
+	client    *openai.Client
+	modelName string
+	request   *openai.ChatCompletionNewParams
+}
+
+// NewModelGenerator creates a new ModelGenerator instance
+func NewModelGenerator(client *openai.Client, modelName string) *ModelGenerator {
+	return &ModelGenerator{
+		client:    client,
+		modelName: modelName,
+		request: &openai.ChatCompletionNewParams{
+			Model: openai.F(modelName),
+		},
+	}
+}
+
+// WithMessages adds messages to the request
+func (g *ModelGenerator) WithMessages(messages []*ai.Message) *ModelGenerator {
+	oaiMessages := make([]openai.ChatCompletionMessageParamUnion, 0, len(messages))
+	for _, msg := range messages {
+		content := g.concatenateContent(msg.Content)
+		switch msg.Role {
+		case ai.RoleSystem:
+			oaiMessages = append(oaiMessages, openai.SystemMessage(content))
+		case ai.RoleModel:
+			oaiMessages = append(oaiMessages, openai.AssistantMessage(content))
+		case ai.RoleTool:
+			oaiMessages = append(oaiMessages, openai.ToolMessage("", content)) // TODO: Add tool ID if needed
+		default:
+			oaiMessages = append(oaiMessages, openai.UserMessage(content))
+		}
+	}
+	g.request.Messages = openai.F(oaiMessages)
+	return g
+}
+
+// WithConfig adds configuration parameters from the model request
+func (g *ModelGenerator) WithConfig(config any) *ModelGenerator {
+	if config != nil {
+		// TODO: Implement configuration from model request
+		// modelRequest.Config is any type
+	}
+	return g
+}
+
+// WithTools adds tools to the request
+func (g *ModelGenerator) WithTools(tools []ai.Tool, choice ai.ToolChoice) *ModelGenerator {
+	// TODO: Implement tools from model request
+	// see vertex ai recent pr here for reference: https://github.com/firebase/genkit/pull/2259
+	return g
+}
+
+// Generate executes the generation request
+func (g *ModelGenerator) Generate(ctx context.Context, handleChunk func(context.Context, *ai.ModelResponseChunk) error) (*ai.ModelResponse, error) {
+	if handleChunk != nil {
+		return g.generateStream(ctx, handleChunk)
+	}
+	return g.generateComplete(ctx)
+}
+
+// concatenateContent concatenates text content into a single string
+func (g *ModelGenerator) concatenateContent(parts []*ai.Part) string {
+	content := ""
+	for _, part := range parts {
+		content += part.Text
+	}
+	return content
+}
+
+// generateStream generates a streaming model response
+func (g *ModelGenerator) generateStream(ctx context.Context, handleChunk func(context.Context, *ai.ModelResponseChunk) error) (*ai.ModelResponse, error) {
+	stream := g.client.Chat.Completions.NewStreaming(ctx, *g.request)
+	defer stream.Close()
+
+	var fullResponse ai.ModelResponse
+	fullResponse.Message = &ai.Message{
+		Role:    ai.RoleModel,
+		Content: make([]*ai.Part, 0),
+	}
+
```

Comment:
Without this change, the test fails due to a nil pointer error.

```suggestion
	fullResponse.Request = &ai.ModelRequest{}

```

Reply:
good catch!

Reply:
nice find - added!